services/api/service.go

```diff
@@ -484,94 +486,144 @@ func (api *RelayAPI) handleRegisterValidator(w http.ResponseWriter, req *http.Re
 		"mevBoostV": common.GetMevBoostVersionFromUserAgent(ua),
 	})
 
-	respondError := func(code int, msg string) {
-		log.Warn("bad request: ", msg)
-		api.RespondError(w, code, msg)
-	}
-
 	start := time.Now()
 	registrationTimeUpperBound := start.Add(10 * time.Second)
 
-	registrations := []types.SignedValidatorRegistration{}
+	numRegTotal := 0
+	numRegProcessed := 0
+	numRegActive := 0
 	numRegNew := 0
+	processingStoppedByError := false
 
-	if err := json.NewDecoder(req.Body).Decode(&registrations); err != nil {
-		respondError(http.StatusBadRequest, "failed to decode payload")
+	respondError := func(code int, msg string) {
+		processingStoppedByError = true
+		log.Warnf("error: %s", msg)
+		api.RespondError(w, code, msg)
+	}
+
+	if req.ContentLength == 0 {
+		respondError(http.StatusBadRequest, "empty request")
+		return
+	}
+
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		log.WithField("contentLength", req.ContentLength).Warn("failed to read request body")
+		api.RespondError(w, http.StatusBadRequest, "failed to read request body")
 		return
 	}
 
-	// Possible optimisations:
-	// - GetValidatorRegistrationTimestamp could keep a cache in memory for some time and check memory first before going to Redis
-	// - Do multiple loops and filter down set of registrations, and batch checks for all registrations instead of locking for each individually:
-	//   (1) sanity checks, (2) IsKnownValidator, (3) CheckTimestamp, (4) Batch SetValidatorRegistration
-	for _, registration := range registrations {
-		if registration.Message == nil {
-			respondError(http.StatusBadRequest, "registration without message")
+	parseRegistration := func(value []byte) (pkHex types.PubkeyHex, timestampInt int64, err error) {
+		pubkey, err := jsonparser.GetUnsafeString(value, "message", "pubkey")
+		if err != nil {
+			return pkHex, timestampInt, fmt.Errorf("registration message error (pubkey): %w", err)
+		}
+
+		timestamp, err := jsonparser.GetUnsafeString(value, "message", "timestamp")
+		if err != nil {
+			return pkHex, timestampInt, fmt.Errorf("registration message error (timestamp): %w", err)
+		}
+
+		timestampInt, err = strconv.ParseInt(timestamp, 10, 64)
+		if err != nil {
+			return pkHex, timestampInt, fmt.Errorf("invalid timestamp: %w", err)
+		}
+
+		return types.PubkeyHex(pubkey), timestampInt, nil
+	}
+
+	// Iterate over the registrations
+	_, err = jsonparser.ArrayEach(body, func(value []byte, dataType jsonparser.ValueType, offset int, _err error) {
```

Comment:
Just want to double check, registration parsing is done sequentially and not in parallel, right? Looking at the library internally, it appears to be the case but I'd appreciate some confirmation. If not, there could be a minor race condition with all of those counter variables. Pretty sure this okay.

Reply:
Good points, and yes it is sequential, and also should be (for avoid DoS vectors and resource exhaustion attacks).